[cppcoro] add new port

luncliff opened the pull request:
Previously there was #7779 to add https://github.com/lewissbaker/cppcoro, but it is closed for some reason.

Since the fork defines minimum CMake configurations, I expect the port will be available in the most Desktop(or so-called non-mobile) triplets.

### Related Issues

* #5106 (port request)

luncliff committed:
[cppcoro] add new port

NancyLi1013 commented on ports/cppcoro/CONTROL:
```diff
@@ -0,0 +1,4 @@
+Source: cppcoro
+Version: 2020.04
+Description: A library of C++ coroutine abstractions for the coroutines TS
```
Could you please add homepage here?
Homepage: https://luncliff.github.io/coroutine/

NancyLi1013 commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+include(vcpkg_common_functions)
```
`include(vcpkg_common_functions) `is no longer needed.
Could you please remove this line?

NancyLi1013 commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+)
+vcpkg_install_cmake()
+
+file(INSTALL     ${SOURCE_PATH}/LICENSE.txt
```
Could you please update 
```
file(INSTALL     ${SOURCE_PATH}/LICENSE.txt
     DESTINATION ${CURRENT_PACKAGES_DIR}/share/cppcoro
     RENAME      copyright
)
```
as
```
file(INSTALL     ${SOURCE_PATH}/LICENSE.txt
     DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
     RENAME      copyright
)
```
?

NancyLi1013 commented:
/azp run

NancyLi1013 commented:
Hi @luncliff 
Thanks for this PR.
Could you please look into the regression on Linux and try to fix it?

luncliff commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+# the repo is fork of https://github.com/lewissbaker/cppcoro to support CMake / VcPkg
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO            luncliff/cppcoro
```
* The original repo doesn't support CMake

The `lewissbaker/cppcoro` already has a PR for CMake support, but it's not merged for a while. I thought I could deal with future port issues until the repo's PR is merged.

What I planned is to change portfile to `lewissbaker/cppcoro` at the moment.

NancyLi1013 replied:
Thanks for your clarification.
We suggest to use original repo instead of the forked one.
Could you please use` lewissbaker/cppcoro` and try to add CMakeLists.txt to _ports/cppcoro/_ folder?

luncliff commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,26 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO            luncliff/cppcoro
+    REF             20b4217a65165d0ec7849fbe7970310cdf31b6a5
```
I will make a tag after linux build check and will update to it.

luncliff commented:
Thanks for very fast review, @NancyLi1013.  
I will apply the requested changes and push again within days.

For https://github.com/microsoft/vcpkg/pull/10693#discussion_r404514478, 
If you think registration of the fork repo is inappropriate, please feel free to let me know.

luncliff commented:
For Linux build failure, the errors are something like this.

```
[1/20] /usr/bin/c++   -I/mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/include -fPIC -g   -std=gnu++17 -MD -MT CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o -MF CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o.d -o CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o -c /mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/lib/async_mutex.cpp
FAILED: CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o 
/usr/bin/c++   -I/mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/include -fPIC -g   -std=gnu++17 -MD -MT CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o -MF CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o.d -o CMakeFiles/cppcoro.dir/lib/async_mutex.cpp.o -c /mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/lib/async_mutex.cpp
In file included from /mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/lib/async_mutex.cpp:6:
/mnt/develop/vcpkg/buildtrees/cppcoro/src/0cdf31b6a5-b830c1617b/include/cppcoro/async_mutex.hpp:8:10: fatal error: experimental/coroutine: No such file or directory
    8 | #include <experimental/coroutine>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
...
```

As far as I know, there are 2 problems.

1. GCC 7 doesn't support C++ Coroutines (TS) feature. We need at least GCC 10
2. The build requires `libc++`, instead of `libstdc++`

For case 1, the user can workaround with `CC=clang`, `CXX=clang`.  
For case 2, triplet (`x64-linux`) customization is required to enforce `libc++`.

luncliff committed:
[cppcoro] update with PR review

* update reference to tag and hash value
* ${PORT} for destination

Linux build support should marked unavailable?

NancyLi1013 commented:
@luncliff 
Thanks for your deep investigation about the regression on Linux.
As far as I'm concerned, most users installed gcc version less than 10.
For `libc++`, we prefer to use the triplet that vcpkg supports instead of custom triplet.
So currently I suggest that we had better disable the support for linux.

NancyLi1013 commented:
/azp run

luncliff committed:
[cppcoro] embed CMakeLists.txt file

* copy the CMakeLists.txt file and start build with it
* change version to 2020.2

For now the repo doesn't have any tags.
Therefore its version followed the latest commit,
which is made in 2020 Feb.

luncliff commented:
Close #5106 

Thanks for your feedback, @NancyLi1013 . I just embedded CMakeLists.txt, and marked `Supports: !(uwp|linux)`.

luncliff commented:
Since the original repo doesn't have tag, I used version code 2020.02. The month is when the latest commit was made.

NancyLi1013 commented:
If there is no released version, you can use a date that you accept the latest commit.
Please refer to this [doc](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/control-files.md#version).

luncliff committed:
[cppcoro] 2020-02-28

luncliff commented:
@NancyLi1013 I will check the other parts again :)

NancyLi1013 commented:
@luncliff Thanks for your update.
Is there anything else you want to check or update now?

luncliff commented:
> @luncliff Thanks for your update.
> Is there anything else you want to check or update now?

Looks good to me for current state. Do I have to rebase the commits?

NancyLi1013 commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,29 @@
+
```
Since this port currently doesn't support uwp and linux, could you please also add
`vcpkg_fail_port_install(ON_TARGET "uwp" "linux")`?

luncliff committed:
[cppcoro] port install failure on platforms

* vcpkg_fail_port_install on uwp/linux

mention PR #10693 so linux/clang users can see the port issues

JackBoosY commented on ports/cppcoro/CMakeLists.txt:
```diff
@@ -0,0 +1,355 @@
+    )
+    set_target_properties(cppcoro
+    PROPERTIES
+        WINDOWS_EXPORT_ALL_SYMBOLS true
```
Why add this option here?
If no symbols are exported from this port, they should be marked as static library in Windows.

luncliff replied:
Thanks for the point @JackBoosY. The option is what I habitually use. I agree that it is not necessary for this case. 
I will remove it in next commit.

JackBoosY commented on ports/cppcoro/CMakeLists.txt:
```diff
@@ -0,0 +1,355 @@
+elseif(MSVC)
+    target_compile_options(cppcoro
+    PUBLIC
+        /std:c++latest /await
```
These options have been added in the vcpkg preprocessing step. So we don't need it here.
Same with linux and osx.

luncliff replied:
Hmm.. I didn't know that... Would you link the part?  
I have another PR which is using these options, so I'd like to make sure of it myself.

JackBoosY replied:
@luncliff See [here](https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/windows.cmake).
If you need a specific C++ version to build this port, please write it here.

luncliff replied:
@JackBoosY, I searched for `/await` option but couldn't find it.
I think I'm not understanding your intend correctly.

Unless `VCPKG_C_FLAGS` or `VCPKG_CXX_FLAGS` is set, the option won't be used.
If so, do I have to alert for the point? 

Or did you mean to move those options to `portfile.cmake`?

JackBoosY replied:
Okay, that's fine.

JackBoosY commented on ports/cppcoro/CMakeLists.txt:
```diff
@@ -0,0 +1,355 @@
+if(WIN32)
+    set_target_properties(cppcoro
+    PROPERTIES
+        WINDOWS_EXPORT_ALL_SYMBOLS True
```
Same as above question.

luncliff committed:
[cppcoro] force static linkage

* report error for clang-cl

build with VC++ in VS 2019 will fail by the header file.

JackBoosY commented on ports/cppcoro/portfile.cmake:
```diff
@@ -0,0 +1,33 @@
+
+# for Linux, the port requires libc++ and Clang. see #10693
+vcpkg_fail_port_install(ON_TARGET "uwp" "linux")
```
Please add the following codes to _VCPKG_PATH/scripts/ci.baseline.txt_:
```
cppcoro:x64-linux=fail
cppcoro:arm-uwp=fail
cppcoro:x64-uwp=fail
```

luncliff committed:
[cppcoro] register ci.baseline failure

NancyLi1013 commented:
/azp run

ras0219-msft committed:
[cppcoro] Simplify CMakeLists.txt to ease future maintenance. Fix VS 2019 builds.

ras0219-msft commented:
Thanks for the PR!

luncliff commented:
Appreciate for the reviews! Would you close the #5106?
